presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExpressionEquivalence.java

```diff
@@ -0,0 +1,307 @@
+        // check for missing coercions
+        ExpressionAnalysis expressionAnalysis = analyzeExpressionsWithSymbols(
+                session,
+                metadata,
+                sqlParser,
+                symbolInput.keySet().stream()
+                        .collect(Collectors.toMap(identity(), symbol -> inputTypes.get(symbolInput.get(symbol)))),
+                ImmutableList.of(expression));
+
+        // add coercions
+        Expression coercedExpression = ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Void>()
+        {
+            @Override
+            public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                Expression rewrittenExpression = treeRewriter.defaultRewrite(node, context);
+
+                // cast expression if coercion is registered
+                Type coerceToType = expressionAnalysis.getCoercion(node);
+
+                if (coerceToType != null) {
+                    rewrittenExpression = new Cast(rewrittenExpression, coerceToType.getTypeSignature().toString());
+                }
+
+                return rewrittenExpression;
+            }
+        }, expression);
+
+        // replace qualified names with input references since row expressions do not support these
+        Expression expressionWithInputReferences = ExpressionTreeRewriter.rewriteWith(new SymbolToInputRewriter(symbolInput), coercedExpression);
+
+        // determine the type of every expression
+        IdentityHashMap<Expression, Type> expressionTypes = getExpressionTypesFromInput(
+                session,
+                metadata,
+                sqlParser,
+                inputTypes,
+                expressionWithInputReferences);
+
+        // convert to row expression
+        return translate(expressionWithInputReferences, SCALAR, expressionTypes, metadata.getFunctionRegistry(), metadata.getTypeManager(), session, true);
+    }
+
+    private static class CanonicalizationVisitor
+            implements RowExpressionVisitor<Void, RowExpression>
+    {
+        @Override
+        public RowExpression visitCall(CallExpression call, Void context)
+        {
+            call = new CallExpression(
+                    call.getSignature(),
+                    call.getType(),
+                    call.getArguments().stream()
+                            .map(expression -> expression.accept(this, context))
+                            .collect(toImmutableList()));
+
+            String callName = call.getSignature().getName();
+
+            if (callName.equals("AND") || callName.equals("OR")) {
+                // return sorted distinct arguments
+                Set<RowExpression> distinctArguments = ImmutableSet.copyOf(call.getArguments());
+                if (distinctArguments.size() == 1) {
+                    return Iterables.getOnlyElement(distinctArguments);
+                }
+
+                return new CallExpression(
```

Comment:
You can do this easily. See: com.facebook.presto.sql.ExpressionUtils#extractConjuncts